proto/cosmos/fee_grant/v1beta1/fee_grant.proto

```diff
@@ -0,0 +1,59 @@
+syntax = "proto3";
+package cosmos.fee_grant.v1beta1;
+
+import "gogoproto/gogo.proto";
+import "google/protobuf/any.proto";
+import "cosmos_proto/cosmos.proto";
+import "cosmos/base/v1beta1/coin.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/cosmos/cosmos-sdk/x/fee_grant/types";
+
+// BasicFeeAllowance implements FeeAllowance with a one-time grant of tokens
+// that optionally expires. The delegatee can use up to SpendLimit to cover fees.
+message BasicFeeAllowance {
+  option (cosmos_proto.implements_interface) = "FeeAllowanceI";
+
+  repeated cosmos.base.v1beta1.Coin spend_limit = 1
+      [(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
+  ExpiresAt expiration = 2 [(gogoproto.nullable) = false];
+}
+
+// PeriodicFeeAllowance extends FeeAllowance to allow for both a maximum cap,
+// as well as a limit per time period.
+message PeriodicFeeAllowance {
+  option (cosmos_proto.implements_interface) = "FeeAllowanceI";
+
+  BasicFeeAllowance basic                              = 1 [(gogoproto.nullable) = false];
+  Duration          period                             = 2 [(gogoproto.nullable) = false];
+  repeated cosmos.base.v1beta1.Coin period_spend_limit = 3
+      [(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
+  repeated cosmos.base.v1beta1.Coin period_can_spend = 4
+      [(gogoproto.nullable) = false, (gogoproto.castrepeated) = "github.com/cosmos/cosmos-sdk/types.Coins"];
+
+  ExpiresAt period_reset = 5 [(gogoproto.nullable) = false];
+}
+
+// Duration is a repeating unit of either clock time or number of blocks.
+// This is designed to be added to an ExpiresAt struct.
+message Duration {
+
+  google.protobuf.Timestamp clock = 1 [(gogoproto.stdduration) = true, (gogoproto.nullable) = false];
+  int64                     block = 2;
+}
+
+// ExpiresAt is a point in time where something expires.
+// It may be *either* block time or block height
+message ExpiresAt {
+
+  google.protobuf.Timestamp time   = 1 [(gogoproto.stdtime) = true, (gogoproto.nullable) = false];
+  int64                     height = 2;
+}
+
+// FeeAllowanceGrant is stored in the KVStore to record a grant with full context
+message FeeAllowanceGrant {
+
+  string              granter   = 1;
+  string              grantee   = 2;
+  google.protobuf.Any allowance = 3 [(cosmos_proto.accepts_interface) = "FeeAllowanceI"];
```

Comment:
why don't we use here `oneof {PeriodicFeeAllowance, BasicFeeAllowance}`? 

Oneof is safe for future extensions, right?